File: extensions/memory-lance-context/src/bridge.ts
import { spawn, type ChildProcess } from "node:child_process";
import { randomUUID } from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import * as readline from "node:readline";
import type {
  AddParams,
  SearchParams,
  ListParams,
  CompactParams,
  CompactionMetrics,
  CompactionStats,
  SearchHit,
  ContextRecord,
} from "./types.js";

type BridgeConfig = {
  pythonPath: string;
  dbPath: string;
  timeout: number;
  compaction?: {
    enabled: boolean;
    check_interval_secs: number;
    min_fragments: number;
    target_rows_per_fragment: number;
    quiet_hours: Array<[number, number]>;
  };
  storage_options?: {
    aws_access_key_id?: string;
    aws_secret_access_key?: string;
    aws_region?: string;
    aws_endpoint_url?: string;
  };
  debug?: boolean;
};

type PendingRequest = {
  resolve: (value: unknown) => void;
  reject: (error: Error) => void;
  timeout: NodeJS.Timeout;
};

type BridgeResponse =
  | { id: string; ok: true; result: unknown }
  | { id: string; ok: false; error: string };

/**
 * Python bridge for lance-context via JSON-over-stdio.
 * Spawns a Python subprocess running bridge.py and communicates via JSON lines.
 */
export class LanceContextBridge {
  private proc: ChildProcess | null = null;
  private pendingRequests = new Map<string, PendingRequest>();
  private readyPromise: Promise<void> | null = null;
  private closed = false;

  constructor(private config: BridgeConfig) {}

  /**
   * Initialize the bridge by spawning the Python subprocess.
   */
  async initialize(): Promise<void> {
    if (this.proc) {
      return;
    }

    this.readyPromise = new Promise((resolve, reject) => {
      // Validate Python path (security: prevent arbitrary executable execution)
      const pythonPath = this.resolvePythonPath(this.config.pythonPath);

      // Locate bridge.py
      const bridgePath = this.resolveBridgePath();

      // Ensure db directory exists
      const dbDir = path.dirname(this.config.dbPath);
      if (!fs.existsSync(dbDir)) {
        fs.mkdirSync(dbDir, { recursive: true });
      }

      // Build config JSON
      const config = {
        compaction: this.config.compaction,
        storage_options: this.config.storage_options,
        debug: this.config.debug,
      };

      // Spawn Python bridge
      const args = [bridgePath, "--uri", this.config.dbPath, "--config", JSON.stringify(config)];

      this.proc = spawn(pythonPath, args, {
        stdio: ["pipe", "pipe", "pipe"],
        env: { ...process.env },
      });

      // Setup JSON line protocol on stdout
      const rl = readline.createInterface({ input: this.proc.stdout! });
      let ready = false;

      rl.on("line", (line) => {
        try {
          const response: BridgeResponse | { ok: true; status: "ready" } = JSON.parse(line);

          // Handle ready signal
          if ("status" in response && response.status === "ready") {
            ready = true;
            resolve();
            return;
          }

          // Handle command response
          if ("id" in response) {
            const pending = this.pendingRequests.get(response.id);
            if (pending) {
              clearTimeout(pending.timeout);
              this.pendingRequests.delete(response.id);

              if (response.ok) {
                pending.resolve(response.result);
              } else {
                pending.reject(new Error(response.error));
              }
            }
          }
        } catch (err) {
          // Ignore malformed lines
        }
      });

      // Capture stderr for debugging
      let stderr = "";
      this.proc.stderr?.setEncoding("utf8");
      this.proc.stderr?.on("data", (chunk) => {
        stderr += String(chunk);
      });

      // Handle process exit
      this.proc.on("exit", (code, signal) => {
        this.proc = null;
        const error = new Error(
          `Python bridge exited (code: ${code}, signal: ${signal})\n${stderr.trim()}`,
        );

        if (!ready) {
          reject(error);
        }

        // Reject all pending requests
        for (const pending of this.pendingRequests.values()) {
          clearTimeout(pending.timeout);
          pending.reject(error);
        }
        this.pendingRequests.clear();
      });

      // Handle spawn errors
      this.proc.on("error", (err) => {
        const wrapped = new Error(`Failed to spawn Python bridge: ${err.message}`);
        if (!ready) {
          reject(wrapped);
        }
      });
    });

    return this.readyPromise;
  }

  /**
   * Send a command to the Python bridge and wait for response.
   */
  private async call<T>(command: string, params: Record<string, unknown> = {}): Promise<T> {
    if (this.closed) {
      throw new Error("Bridge is closed");
    }

    await this.readyPromise;

    if (!this.proc || !this.proc.stdin) {
      throw new Error("Bridge not initialized");
    }

    const requestId = randomUUID();
    const request = { id: requestId, command, params };

    // Write JSON line to stdin
    this.proc.stdin.write(JSON.stringify(request) + "\n");

    // Wait for response with timeout
    return new Promise<T>((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pendingRequests.delete(requestId);
        reject(new Error(`Bridge timeout: ${command} (${this.config.timeout}ms)`));
      }, this.config.timeout);

      this.pendingRequests.set(requestId, {
        resolve: resolve as (value: unknown) => void,
        reject,
        timeout,
      });
    });
  }

  /**
   * Add a record to the context.
   */
  async add(params: AddParams): Promise<void> {
    await this.call<void>("add", params);
  }

  /**
   * Search the context using semantic similarity.
   */
  async search(params: SearchParams): Promise<SearchHit[]> {
    const results = await this.call<SearchHit[]>("search", params);
    return results;
  }

  /**
   * List records chronologically.
   */
  async list(params: ListParams = {}): Promise<ContextRecord[]> {
    const results = await this.call<ContextRecord[]>("list", params);
    return results;
  }

  /**
   * Create a named snapshot.
   */
  async snapshot(label?: string): Promise<string> {
    const result = await this.call<{ snapshot_id: string }>("snapshot", { label });
    return result.snapshot_id;
  }

  /**
   * Checkout to a specific version or snapshot.
   */
  async checkout(versionId: number | string): Promise<void> {
    await this.call<void>("checkout", { version_id: versionId });
  }

  /**
   * Manually trigger compaction.
   */
  async compact(params?: CompactParams): Promise<CompactionMetrics> {
    const result = await this.call<CompactionMetrics>("compact", params || {});
    return result;
  }

  /**
   * Get compaction statistics.
   */
  async compactionStats(): Promise<CompactionStats> {
    const result = await this.call<CompactionStats>("compaction_stats");
    return result;
  }

  /**
   * Get the number of entries.
   */
  async entries(): Promise<number> {
    const result = await this.call<{ count: number }>("entries");
    return result.count;
  }

  /**
   * Get the current version.
   */
  async version(): Promise<number> {
    const result = await this.call<{ version: number }>("version");
    return result.version;
  }

  /**
   * Get the dataset URI.
   */
  async uri(): Promise<string> {
    const result = await this.call<{ uri: string }>("uri");
    return result.uri;
  }

  /**
   * Get the current branch.
   */
  async branch(): Promise<string> {
    const result = await this.call<{ branch: string }>("branch");
    return result.branch;
  }

  /**
   * Close the bridge gracefully.
   */
  async close(): Promise<void> {
    if (this.closed || !this.proc) {
      return;
    }

    this.closed = true;

    // Close stdin to signal Python to exit gracefully
    if (this.proc.stdin) {
      this.proc.stdin.end();
    }

    // Wait for process to exit (with timeout)
    await new Promise<void>((resolve) => {
      if (!this.proc) {
        resolve();
        return;
      }

      const timeout = setTimeout(() => {
        if (this.proc) {
          this.proc.kill("SIGTERM");
        }
        resolve();
      }, 5000);

      this.proc.once("exit", () => {
        clearTimeout(timeout);
        resolve();
      });
    });
  }

  /**
   * Resolve and validate Python executable path.
   */
  private resolvePythonPath(pythonPath: string): string {
    // SECURITY: Prevent arbitrary executable execution
    const base = path.basename(pythonPath).toLowerCase();
    const allowed = ["python", "python3", "python3.exe", "python.exe"];

    if (!allowed.some((name) => base === name || base.startsWith(name + "."))) {
      throw new Error(`Invalid Python executable: ${pythonPath}`);
    }

    // If absolute path, validate it exists
    if (path.isAbsolute(pythonPath)) {
      if (!fs.existsSync(pythonPath)) {
        throw new Error(`Python executable not found: ${pythonPath}`);
      }
    }

    return pythonPath;
  }

  /**
   * Resolve bridge.py path.
   */
  private resolveBridgePath(): string {
    // Try relative to this file
    const candidates = [
      path.join(__dirname, "..", "python", "bridge.py"),
      path.join(process.cwd(), "extensions", "memory-lance-context", "python", "bridge.py"),
    ];

    for (const candidate of candidates) {
      if (fs.existsSync(candidate)) {
        return candidate;
      }
    }

    throw new Error(
      `bridge.py not found. Searched: ${candidates.join(", ")}\n` +
        "Please ensure extensions/memory-lance-context/python/bridge.py exists.",
    );
  }
}

File: extensions/memory-lance-context/python/bridge.py
#!/usr/bin/env python3
"""
JSON-over-stdio bridge for lance-context.
Reads JSON commands from stdin, writes JSON responses to stdout.

Protocol:
  Input: JSON lines with {"id": str, "command": str, "params": dict}
  Output: JSON lines with {"id": str, "ok": bool, "result": any} or {"id": str, "ok": bool, "error": str}
"""
import sys
import json
import traceback
from typing import Any

try:
    from lance_context.api import Context
except ImportError as exc:
    print(json.dumps({"ok": False, "error": f"Failed to import lance_context: {exc}"}), flush=True)
    sys.exit(1)


def serialize_result(result: Any) -> Any:
    """Convert Python objects to JSON-serializable types."""
    if result is None:
        return None
    if isinstance(result, (str, int, float, bool)):
        return result
    if isinstance(result, dict):
        return {k: serialize_result(v) for k, v in result.items()}
    if isinstance(result, (list, tuple)):
        return [serialize_result(item) for item in result]
    if hasattr(result, "isoformat"):  # datetime
        return result.isoformat()
    if isinstance(result, bytes):
        # Convert binary to base64 for JSON serialization
        import base64
        return base64.b64encode(result).decode("ascii")
    return str(result)


def handle_command(ctx: Context, cmd: dict[str, Any]) -> Any:
    """Handle a single command and return the result."""
    command = cmd["command"]
    params = cmd.get("params", {})

    if command == "add":
        ctx.add(
            role=params["role"],
            content=params["content"],
            embedding=params.get("embedding"),
            bot_id=params.get("bot_id"),
            session_id=params.get("session_id"),
        )
        return None

    elif command == "search":
        results = ctx.search(params["query"], params.get("limit"))
        return results

    elif command == "list":
        results = ctx.list(params.get("limit"), params.get("offset"))
        return results

    elif command == "snapshot":
        snapshot_id = ctx.snapshot(params.get("label"))
        return {"snapshot_id": snapshot_id}

    elif command == "checkout":
        version_id = params["version_id"]
        if isinstance(version_id, str):
            try:
                version_id = int(version_id)
            except ValueError:
                pass  # Keep as string for named snapshots
        ctx.checkout(version_id)
        return None

    elif command == "compact":
        metrics = ctx.compact(
            target_rows_per_fragment=params.get("target_rows_per_fragment"),
            materialize_deletions=params.get("materialize_deletions", True),
        )
        return metrics

    elif command == "compaction_stats":
        stats = ctx.compaction_stats()
        return stats

    elif command == "entries":
        return {"count": ctx.entries()}

    elif command == "version":
        return {"version": ctx.version()}

    elif command == "uri":
        return {"uri": ctx.uri()}

    elif command == "branch":
        return {"branch": ctx.branch()}

    else:
        raise ValueError(f"Unknown command: {command}")


def main():
    """Main loop: read commands from stdin, write responses to stdout."""
    import argparse
    parser = argparse.ArgumentParser(description="lance-context JSON bridge")
    parser.add_argument("--uri", required=True, help="Lance dataset URI")
    parser.add_argument("--config", type=json.loads, default="{}", help="JSON config")
    args = parser.parse_args()

    config = args.config

    # Extract compaction config
    compaction = config.get("compaction", {})
    storage_options = config.get("storage_options", {})

    # Create Context with compaction config
    try:
        ctx = Context.create(
            args.uri,
            storage_options=storage_options if storage_options else None,
            enable_background_compaction=compaction.get("enabled", False),
            compaction_interval_secs=compaction.get("check_interval_secs", 300),
            compaction_min_fragments=compaction.get("min_fragments", 5),
            compaction_target_rows=compaction.get("target_rows_per_fragment", 1_000_000),
            quiet_hours=compaction.get("quiet_hours", []),
        )
    except Exception as exc:
        error_msg = f"Failed to create Context: {exc}\n{traceback.format_exc()}"
        print(json.dumps({"ok": False, "error": error_msg}), flush=True)
        sys.exit(1)

    # Signal ready
    print(json.dumps({"ok": True, "status": "ready"}), flush=True)

    # Command loop: read JSON lines from stdin
    for line in sys.stdin:
        try:
            cmd = json.loads(line)
            result = handle_command(ctx, cmd)
            serialized = serialize_result(result)
            response = {"id": cmd["id"], "ok": True, "result": serialized}
        except Exception as exc:
            error_msg = f"{type(exc).__name__}: {exc}"
            # Include traceback for debugging (can be disabled in production)
            if config.get("debug"):
                error_msg += f"\n{traceback.format_exc()}"
            response = {"id": cmd.get("id"), "ok": False, "error": error_msg}

        print(json.dumps(response), flush=True)


if __name__ == "__main__":
    main()
